lib/core/figbird.js

```diff
@@ -0,0 +1,604 @@
+    const serviceName = this.#serviceNamesByQueryId.get(queryId)
+    if (serviceName) {
+      const service = this.getState().get(serviceName)
+      if (service) {
+        return service.queries.get(queryId)
+      }
+    }
+  }
+
+  getQueryState(queryId) {
+    return this.#getQuery(queryId)?.state
+  }
+
+  materialize(queryRef) {
+    const { queryId, desc, config } = queryRef.details()
+
+    if (!this.#getQuery(queryId)) {
+      this.#serviceNamesByQueryId.set(queryId, desc.serviceName)
+
+      this.#transactOverService(
+        queryId,
+        service => {
+          service.queries.set(queryId, {
+            queryId,
+            desc,
+            config,
+            pending: true,
+            filterItem: this.#createItemFilter(desc, config),
+            state: config.skip
+              ? {
+                  data: null,
+                  status: 'idle',
+                  isFetching: false,
+                  error: null,
+                }
+              : {
+                  data: null,
+                  status: 'loading',
+                  isFetching: true,
+                  error: null,
+                },
+          })
+        },
+        { silent: true },
+      )
+    }
+  }
+
+  #createItemFilter(desc, config) {
+    const filterItem = this.#adapter.matcher(desc.params?.query)
+    return config.matcher ? config.matcher(desc.params?.query, filterItem) : filterItem
+  }
+
+  #addListener(queryId, fn) {
+    if (!this.#listeners.has(queryId)) {
+      this.#listeners.set(queryId, new Set())
+    }
+    this.#listeners.get(queryId).add(fn)
+    return () => {
+      const listeners = this.#listeners.get(queryId)
+      if (listeners) {
+        listeners.delete(fn)
+        if (listeners.size === 0) {
+          this.#listeners.delete(queryId)
+        }
+      }
+    }
+  }
+
+  #invokeListeners(queryId) {
+    const listeners = this.#listeners.get(queryId)
+    if (listeners) {
+      const state = this.getQueryState(queryId)
+      listeners.forEach(listener => listener(state))
+    }
+  }
+
+  #addGlobalListener(fn) {
+    this.#globalListeners.add(fn)
+    return () => {
+      this.#globalListeners.delete(fn)
+    }
+  }
+
+  #invokeGlobalListeners() {
+    const state = this.getState()
+    this.#globalListeners.forEach(listener => listener(state))
+  }
+
+  #listenerCount(queryId) {
+    return this.#listeners.get(queryId)?.size || 0
+  }
+
+  subscribe(queryId, fn) {
+    const q = this.#getQuery(queryId)
+    if (
+      q.pending ||
+      (q.state.status === 'success' && q.config.fetchPolicy === 'swr' && !q.state.isFetching) ||
+      (q.state.status === 'error' && !q.state.isFetching)
+    ) {
+      this.#queue(queryId)
+    }
+
+    const removeListener = this.#addListener(queryId, fn)
+
+    this.#subscribeToRealtime(queryId)
+
+    const shouldVacuumByDefault = q.config.fetchPolicy === 'network-only'
+    return ({ vacuum = shouldVacuumByDefault } = {}) => {
+      removeListener()
```

Comment:
I think maybe here we should unsubscribe realtime too?

Reply:
But then queries can get stale, for no good reason? Why not keep integrating those events if the websocket is receiving them?

Reply:
I see. But if the unsubscribe actually only stopped the listening when all queries have torn down then I think it makes sense, as there would be no queries using them?

Maybe at the point vacuum runs below, as that's clearing out the query anyway so it will need to be fetched again next time it's used?